providers/class-two-factor-provider.php

```diff
@@ -72,6 +72,43 @@ public function pre_process_authentication( $user ) {
 	 */
 	abstract public function validate_authentication( $user );
 
+	/**
+	 * Logs the failed authentication.
+	 *
+	 * @param WP_User      $user WP_User object of the user trying to login.
+	 * @param string|false $code The code used to authenticate, if available.
+	 *
+	 * @since 0.9.0
+	 *
+	 * @return void
+	 */
+	public function log_failure( $user, $code = false ) {
+		/**
+		 * This action is triggered when a Two Factor validation fails.
+		 *
+		 * @param WP_User             $user WP_User object of the user trying to login.
+		 * @param string|false        $code The code used to authenticate, if available.
+		 * @param Two_Factor_Provider $this The Provider class used during the authentication.
+		 */
+		do_action( 'two_factor_user_login_failed', $user, $code, $this );
+
+		/* translators: %1$d: the user's ID %2$s: the code used to authenticate */
+		$log_message = sprintf( esc_html__( 'The user with ID %1$d failed to login using the code "%2$s"', 'two-factor' ), $user->ID, esc_html( $code ) );
```

Comment:
Would username be better here? Otherwise it's not very informative unless the recepient has IDs memorized, or manually look it up.

Reply:
Makes sense! I just added it [here](https://github.com/WordPress/two-factor/pull/462/commits/bf6bd6a8d5ef5b9ed5f8d31b20d7c9bddf086cba)

Reply:
Nice 👍🏻